Failure when removing design documents (for indexes)

Cloudant 2.4.1
Using a previously suggested method to list design documents : 
```java
List<String> designDocIds = database.getAllDocsRequestBuilder()
                                    .startKey("_design/")
                                    .endKey("_design0")
                                    .build()
                                    .getResponse()
                                    .getDocIds();
```

I then remove each one using : 
`DesignDocumentManager ddm = database.getDesignDocumentManager();
ddm.remove(designDocId);`

This works fine for views etc, but when it hits a DD for Indexes, I get this error : 
```
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected a string but was BEGIN_OBJECT at line 1 column 133
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:176)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:40)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:187)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:93)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:172)
	at com.google.gson.Gson.fromJson(Gson.java:803)
	at com.google.gson.Gson.fromJson(Gson.java:741)
	at com.cloudant.client.org.lightcouch.internal.CouchDbUtil.getResponse(CouchDbUtil.java:211)
	at com.cloudant.client.org.lightcouch.CouchDbClient.get(CouchDbClient.java:329)
	at com.cloudant.client.org.lightcouch.CouchDatabaseBase.find(CouchDatabaseBase.java:88)
	at com.cloudant.client.api.Database.find(Database.java:536)
	at com.cloudant.client.api.DesignDocumentManager.remove(DesignDocumentManager.java:154)```

Hi,

Here is the method I call to remove all design docs, but leave the "data" in place, used for migrating views etc. I added a printStackTrace() to illustrate the error, and include the trace and design doc below.
```java
public void removeDesignDocs(Database database) {
		try {
			List<String> designDocIds = database.getAllDocsRequestBuilder()
					.startKey("_design/")
					.endKey("_design0")
					.includeDocs(true)
					.build()
					.getResponse()
					.getDocIds();
			DesignDocumentManager ddm = database.getDesignDocumentManager();
			for (String id : designDocIds) {
				try {
					log.info("Removing design doc {}",id);
					ddm.remove(id);
				}
				catch (Throwable t) {
					t.printStackTrace();
					try {
						JsonObject jo = database.find(JsonObject.class, id);
						log.info("Removing index {}",id);
						database.remove(id, jo.get("_rev").getAsString());
					}
					catch (Throwable t2) {
						log.error("Problem removing index",t2);
					}
				}
			}
		}
		catch (Throwable t) {
			log.error("Problem removing design docs from db",t);
		}
	}
```
The stack trace : 
```
11-04-16 22:06:43 INFO  WsClient(217): Removing design doc _design/tenantUidModifiedAsc
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected a string but was BEGIN_OBJECT at line 1 column 132
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:176)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:40)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:187)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:93)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:172)
	at com.google.gson.Gson.fromJson(Gson.java:803)
	at com.google.gson.Gson.fromJson(Gson.java:741)
	at com.cloudant.client.org.lightcouch.internal.CouchDbUtil.getResponse(CouchDbUtil.java:211)
	at com.cloudant.client.org.lightcouch.CouchDbClient.get(CouchDbClient.java:329)
	at com.cloudant.client.org.lightcouch.CouchDatabaseBase.find(CouchDatabaseBase.java:88)
	at com.cloudant.client.api.Database.find(Database.java:536)
	at com.cloudant.client.api.DesignDocumentManager.remove(DesignDocumentManager.java:154)
	at com.xxx.sw.project.WsClient.removeDesignDocs(WsClient.java:218)
	at com.xxx.sw.project.WsClient.updateDesignDoc(WsClient.java:281)
	at com.xxx.sw.project.WsClient.upgrade(WsClient.java:239)
	at com.xxx.sw.project.WsClient.connect(WsClient.java:134)
	at com.xxx.sw.project.WsClient.<init>(WsClient.java:86)
	at com.xxx.sw.project.WsClient.getInstance(WsClient.java:92)
	at com.xxx.sw.project.BaseTest.setUp(BaseTest.java:36)
	at com.xxx.sw.project.TestWsClientNoPopulation.setUp(TestWsClientNoPopulation.java:53)
	at junit.framework.TestCase.runBare(TestCase.java:139)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
Caused by: java.lang.IllegalStateException: Expected a string but was BEGIN_OBJECT at line 1 column 132
	at com.google.gson.stream.JsonReader.nextString(JsonReader.java:821)
	at com.google.gson.internal.bind.TypeAdapters$13.read(TypeAdapters.java:358)
	at com.google.gson.internal.bind.TypeAdapters$13.read(TypeAdapters.java:346)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:93)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:172)
	... 34 more
```

The _design/tenantUidModifiedAsc. Oddly I never set the reduce function - it puts that in by default for some reason. : 
```javascript
{
  "_id": "_design/tenantUidModifiedAsc",
  "_rev": "23-f8817c7aecb1789198df53cfe878ff95",
  "language": "query",
  "views": {
    "cardIndex": {
      "map": {
        "fields": {
          "owner.tenantUid": "asc",
          "modified": "asc"
        }
      },
      "reduce": "_count",
      "options": {
        "def": {
          "fields": [
            {
              "owner.tenantUid": "asc"
            },
            {
              "modified": "asc"
            }
          ]
        }
      }
    }
  }
}
```

Let me know if you need anything else.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,8 @@
 - [NEW] Added an optional `SettableViewParameters.STALE_NO` constant for the default omitted case of
   the stale parameter on a view request.
 - [NEW] Added `descending` option for changes feed.
+- [FIX] `JsonSyntaxException` when deserializing Cloudant query language generated design
+  documents into the `DesignDocument` class.
 
 # 2.4.3 (2016-05-05)
 - [IMPROVED] Reduced the length of the User-Agent header string.

--- a/cloudant-client/src/main/java/com/cloudant/client/api/model/DesignDocument.java
+++ b/cloudant-client/src/main/java/com/cloudant/client/api/model/DesignDocument.java
@@ -15,7 +15,9 @@
 
 package com.cloudant.client.api.model;
 
+import com.google.gson.Gson;
 import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.annotations.SerializedName;
 
@@ -55,6 +57,7 @@ import java.util.Map;
  * Map<String, String> updates = new HashMap<String, String>();
  * updates.put("newUpdateHandler", "function (doc, req) { ... }");
  * ddoc.setUpdates(updates);
+ * // Update the design document
  * database.update(ddoc);
  *
  * }
@@ -65,6 +68,10 @@ import java.util.Map;
  * @since 0.0.2
  */
 public class DesignDocument extends com.cloudant.client.org.lightcouch.Document {
+
+    private static final String LANG_QUERY = "query";
+    // Default GSON instance for serializing/deserializing JsonElements of views
+    private static final Gson GSON = new Gson();
 
     private String language;
     private Map<String, MapReduce> views;
@@ -383,9 +390,13 @@ public class DesignDocument extends com.cloudant.client.org.lightcouch.Document 
      * @author Ahmed Yehia
      */
     public static class MapReduce {
-        private String map;
-        private String reduce;
+        private JsonElement map;
+        private JsonElement reduce;
         private String dbcopy;
+
+        public MapReduce() {
+            super();
+        }
 
         /**
          * Get the defined map function.
@@ -393,7 +404,7 @@ public class DesignDocument extends com.cloudant.client.org.lightcouch.Document 
          * @return string of the javascript map function
          */
         public String getMap() {
-            return map;
+            return stringifyMRElement(map);
         }
 
         /**
@@ -402,7 +413,7 @@ public class DesignDocument extends com.cloudant.client.org.lightcouch.Document 
          * @return string of the javascript reduce function
          */
         public String getReduce() {
-            return reduce;
+            return stringifyMRElement(reduce);
         }
 
         /**
@@ -411,7 +422,7 @@ public class DesignDocument extends com.cloudant.client.org.lightcouch.Document 
          * @param map string of the javascript map function
          */
         public void setMap(String map) {
-            this.map = map;
+            this.map = GSON.toJsonTree(map);
         }
 
         /**
@@ -420,7 +431,7 @@ public class DesignDocument extends com.cloudant.client.org.lightcouch.Document 
          * @param reduce string of the javascript reduce function
          */
         public void setReduce(String reduce) {
-            this.reduce = reduce;
+            this.reduce = GSON.toJsonTree(reduce);
         }
 
         /**
@@ -444,6 +455,25 @@ public class DesignDocument extends com.cloudant.client.org.lightcouch.Document 
          */
         public String getDbCopy() {
             return dbcopy;
+        }
+
+        /**
+         * @param element JSON object or string
+         * @return a string form of the object or the JSON string with leading and trailing "
+         * stripped
+         */
+        private String stringifyMRElement(JsonElement element) {
+            if (element != null) {
+                String stringifiedElement = element.toString();
+                if (element.isJsonPrimitive() && element.getAsJsonPrimitive().isString()) {
+                    // If the element is a JSON string, strip the leading and trailing "
+                    return stringifiedElement.substring(1, stringifiedElement.length() - 1);
+                } else {
+                    return stringifiedElement;
+                }
+            } else {
+                return null;
+            }
         }
 
         @Override

--- a/cloudant-client/src/test/java/com/cloudant/tests/DesignDocumentsTest.java
+++ b/cloudant-client/src/test/java/com/cloudant/tests/DesignDocumentsTest.java
@@ -16,8 +16,10 @@ package com.cloudant.tests;
 
 import static org.hamcrest.CoreMatchers.not;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
 
 import com.cloudant.client.api.CloudantClient;
 import com.cloudant.client.api.Database;
@@ -48,6 +50,7 @@ import org.junit.experimental.categories.Category;
 import java.io.File;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
@@ -401,5 +404,116 @@ public class DesignDocumentsTest {
 
         assertEquals("The retrieved list of design documents should match the expected list",
                 designDocs, designManager.list());
+    }
+
+    /**
+     * Test that for javascript language MR view (i.e. the map function is a JSON string) that the
+     * map function is correctly deserialized as a Java String.
+     * The javascript string must be enclosed in \", for compatibility with JSON objects that might
+     * be stored in maps we have to internally switch between the two, but the library did not
+     * originally deal with both cases so for backwards compatibility must return the JSON strings
+     * without the leading and trailing " when they are java strings.
+     *
+     * @throws Exception
+     * @see #serializeJavascriptView()
+     * @see #serializeQueryDesignDoc()
+     * @see #deserializeQueryDesignDoc()
+     */
+    @Test
+    public void deserializeJavascriptView() throws Exception {
+        DesignDocument queryDDoc = fileToDesignDocument("example");
+        Map<String, DesignDocument.MapReduce> views = queryDDoc.getViews();
+        for (DesignDocument.MapReduce mrView : views.values()) {
+            assertFalse("The map function should not start with \"", mrView.getMap().startsWith
+                    ("\""));
+            assertFalse("The map function should not end with \"", mrView.getMap().endsWith("\""));
+        }
+    }
+
+    /**
+     * Test that when setting a javascript map function it is correctly serialized and deserialized.
+     *
+     * @throws Exception
+     * @see #deserializeJavascriptView()
+     * @see #serializeQueryDesignDoc()
+     * @see #deserializeQueryDesignDoc()
+     */
+    @Test
+    public void serializeJavascriptView() throws Exception {
+        // Create and write a design document with a javascript lang map function
+        String testDDocName = "testJSMapFn";
+        String mapFunction = "function(doc){emit([doc.contentArray[0].boolean,doc.contentArray[0]" +
+                ".creator,doc.contentArray[0].created],doc);}";
+        DesignDocument ddoc = new DesignDocument();
+        ddoc.setId(testDDocName);
+        Map<String, DesignDocument.MapReduce> views = new HashMap<String, DesignDocument
+                .MapReduce>();
+        DesignDocument.MapReduce mr = new DesignDocument.MapReduce();
+        mr.setMap(mapFunction);
+        mr.setReduce("_count");
+        views.put("testView", mr);
+        ddoc.setViews(views);
+        Response r = designManager.put(ddoc);
+
+        // Retrieve the doc and check that the javascript function is correct
+        DesignDocument retrievedDDoc = designManager.get(testDDocName, r.getRev());
+        assertNotNull("There should be a retrieved design doc", retrievedDDoc);
+        Map<String, DesignDocument.MapReduce> retrievedViews = retrievedDDoc.getViews();
+        assertNotNull("There should be views defined on the design doc", retrievedViews);
+        DesignDocument.MapReduce mrView = retrievedViews.get("testView");
+        assertNotNull("There should be a testView in the retrieved design doc", mrView);
+        assertEquals("The map function string should be the expected string",
+                mapFunction, mrView.getMap());
+    }
+
+    /**
+     * Test that if a query language design document is serialized the map function is a string
+     * form of the JSON object.
+     *
+     * @throws Exception
+     * @see #serializeJavascriptView()
+     * @see #deserializeJavascriptView()
+     * @see #deserializeQueryDesignDoc()
+     */
+    @Test
+    public void serializeQueryDesignDoc() throws Exception {
+        DesignDocument queryDDoc = fileToDesignDocument("query");
+        Map<String, DesignDocument.MapReduce> views = queryDDoc.getViews();
+        assertEquals("There should be one view", 1, views.size());
+        for (DesignDocument.MapReduce mrView : views.values()) {
+            assertTrue("The map function should be a javascript function in a JSON form, " +
+                    "so start with {", mrView.getMap().startsWith("{"));
+            assertTrue("The map function should be a javascript function in a JSON form, " +
+                    "so end with }", mrView.getMap().endsWith("}"));
+            assertEquals("The map function string should be an object form",
+                    "{\"fields\":{\"Person_dob\":\"asc\"}}", mrView.getMap());
+        }
+    }
+
+    /**
+     * Test that deserializing a query language design document results in the correct form of the
+     * map function.
+     *
+     * @throws Exception
+     * @see #serializeJavascriptView()
+     * @see #deserializeJavascriptView()
+     * @see #serializeQueryDesignDoc()
+     */
+    @Test
+    public void deserializeQueryDesignDoc() throws Exception {
+        // Put the query design document
+        designManager.put(fileToDesignDocument("query"));
+        // Get the query design document
+        DesignDocument queryDDoc = designManager.get("testQuery");
+        Map<String, DesignDocument.MapReduce> views = queryDDoc.getViews();
+        assertEquals("There should be one view", 1, views.size());
+        for (DesignDocument.MapReduce mrView : views.values()) {
+            assertTrue("The map function should be a javascript function in a JSON form, " +
+                    "so start with {", mrView.getMap().startsWith("{"));
+            assertTrue("The map function should be a javascript function in a JSON form, " +
+                    "so end with }", mrView.getMap().endsWith("}"));
+            assertEquals("The map function string should be an object form",
+                    "{\"fields\":{\"Person_dob\":\"asc\"}}", mrView.getMap());
+        }
     }
 }

--- a/cloudant-client/src/test/resources/design-files/query_design_doc.js
+++ b/cloudant-client/src/test/resources/design-files/query_design_doc.js
@@ -1,0 +1,17 @@
+{
+ "_id": "_design/testQuery",
+ "language": "query",
+ "views": {
+  "testView": {
+   "map": {"fields": {"Person_dob": "asc"}},
+   "reduce": "_count",
+   "options": {
+    "def": {
+     "fields": [
+      "Person_dob"
+     ]
+    }
+   }
+  }
+ }
+}

